src/pages/index.js

```diff
@@ -81,35 +143,174 @@ profileEditButton.addEventListener("click", () => {
 const imagePopup = new PopupWithImage("#preview-modal");
 imagePopup.setEventListeners();
 
-// Enable form validation
+// add event listener to avatar
+document
+  .querySelector(".profile__edit-avatar-button")
+  .addEventListener("click", () => {
+    editAvatarModal.open();
+  });
+// when it fires, open avatar image modal
+
+/* ------------------------ Enable form validation ------------------------ */
+
+const editFormValidator = new FormValidator(
+  validationSettings,
+  profileEditForm
+);
+const addFormValidator = new FormValidator(
+  validationSettings,
+  addCardFormElement
+);
+
+const editAvatarValidator = new FormValidator(
+  validationSettings,
+  document.querySelector("#edit-avatar-modal")
+);
 
 editFormValidator.enableValidation();
 addFormValidator.enableValidation();
-
+editAvatarValidator.enableValidation();
 /* -------------------------------------------------------------------------- */
 /*                                 Functions;                                 */
 /* -------------------------------------------------------------------------- */
-function handleImageClick(data) {
-  imagePopup.open(data);
+function handleCardPreview(cardData) {
+  previewImagePopup.open(cardData);
 }
+
+const avatarSubmitButton = document.querySelector("#avatar-submit-button");
+
+// function handleAvatarEditSubmit(input) {
+//   const link = input.avatar;
+//   api.updateAvatar(link);
+//   if (link) {
+//     avatarSubmitButton.textContent = "Saving...";
+//     api
+
+//       .editProfileImage(link)
+//       .then((userData) => {
+//         userInfo.setUserInfo(userData);
+//         editAvatarModal.close();
+//       })
+//       .catch((err) => console.error("Error updating avatar:", err))
+//       .finally(() => {
+//         editAvatarModal.setLoadingState(false);
+//       });
+//   } else {
+//     console.error("Avatar Link is not defined");
+//   }
+// }
+
+function handleAvatarEditSubmit(input) {
+  const link = input.avatar;
+
+  if (link) {
+    // set loading state to true
+    editAvatarModal.setLoadingState(true);
+    // the right api call is updateAvatar
+    api
+      .updateAvatar(link)
+      .then((userData) => {
+        // set avatar info
+        userInfo.setAvatarInfo(userData.avatar);
+
+        editAvatarModal.close();
+      })
+      .catch((err) => console.error("Error updating avatar:", err))
+      .finally(() => {
+        // set loading state to false
+        editAvatarModal.setLoadingState(false);
```

Comment:
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```